### PR TITLE
pd: support paddle backend and `water/se_e2_a`

### PR DESCRIPTION
**Pull request by HydrogenSulfate**
Split <https://github.com/deepmodeling/deepmd-kit/pull/4157> into several pull requests.

1. Add core modules of paddle backend(`deepmd.pd.*`) and related backend module unitests.
2. Support training/testing/freeze(C++ inference will be supported in subsequent pull request) for example water/se_e2_a.
3. Add se_e2_a related uinttests

Related PR to be merged:

- [x] <https://github.com/PaddlePaddle/Paddle/pull/69139>

### pytorch
image

### paddle:
``` log
deepmd.utils.batch_size                       Adjust batch size from 1024 to 2048
deepmd.utils.batch_size                       Adjust batch size from 2048 to 4096
deepmd.entrypoints.test                       # number of test data : 30 ,
deepmd.entrypoints.test                       Energy MAE         : 7.467160e-02 eV
deepmd.entrypoints.test                       Energy RMSE        : 8.981154e-02 eV
deepmd.entrypoints.test                       Energy MAE/Natoms  : 3.889146e-04 eV
deepmd.entrypoints.test                       Energy RMSE/Natoms : 4.677685e-04 eV
deepmd.entrypoints.test                       Force  MAE         : 4.495974e-02 eV/A
deepmd.entrypoints.test                       Force  RMSE        : 5.883696e-02 eV/A
deepmd.entrypoints.test                       Virial MAE         : 4.683873e+00 eV
deepmd.entrypoints.test                       Virial RMSE        : 6.298489e+00 eV
deepmd.entrypoints.test                       Virial MAE/Natoms  : 2.439517e-02 eV
deepmd.entrypoints.test                       Virial RMSE/Natoms : 3.280463e-02 eV
```

## Release Notes

- **New Features**
	- Introduced support for PaddlePaddle in the DeePMD framework, enhancing model training and evaluation capabilities.
	- Added new backend options and configuration files for multitask models.
	- Implemented new classes and methods for handling Paddle-specific functionalities, including descriptor calculations and model evaluations.
	- Enhanced the command-line interface to include Paddle as a backend option.
	- Expanded the functionality for managing Paddle dependencies and configurations in the testing framework.

- **Bug Fixes**
	- Improved error handling and robustness in various components across the framework.

- **Tests**
	- Expanded the test suite to include Paddle-specific tests, ensuring consistency and reliability across different backends.
	- Introduced unit tests for new functionalities related to Paddle, including model evaluations and descriptor calculations.
	- Added tests to validate force gradient calculations and smoothness properties in models.
	- Implemented tests for neighbor statistics and region transformations, ensuring accuracy in calculations.

- **Documentation**
	- Updated documentation across multiple modules to reflect new features and usage instructions.

### CODE REVIEW

**Comment by HydrogenSulfate** [comment]
Hello @njzjz , I have a question about the naming of the functions `serialize_from_file` and `deserialize_to_file`. According to the docstring explanations for these two functions and the general meaning of "serialize," why can't these functions be named `serialize_to_file` and `deserialize_from_file` instead? I was confused when I first saw these two functions.

**Comment by njzjz** [comment]
> Hello @njzjz , I have a question about the naming of the functions `serialize_from_file` and `deserialize_to_file`. According to the docstring explanations for these two functions and the general meaning of "serialize," why can't these functions be named `serialize_to_file` and `deserialize_from_file` instead? I was confused when I first saw these two functions.

Do you mean `serialize_hook` and `deserialize_hook`?

https://insiders.vscode.dev/github/deepmodeling/deepmd-kit/blob/devel/deepmd/backend/backend.py#L179-L202

**Comment by njzjz** [comment]
> Attention: Patch coverage is `55.25965%` with `2016 lines` in your changes missing coverage. Please review.

About 2000+ lines have yet to be tested in the CI. Could you take a look at the coverage report?

File: deepmd/pd/utils/ase_calc.py
Comment (njzjz) [inline deepmd/pd/utils/ase_calc.py:1]: A simple redirect can be removed. The PyTorch backend has it due to historical reasons, but for a new backend, it is not required.

Comment (HydrogenSulfate) [reply deepmd/pd/utils/ase_calc.py:1]: > A simple redirect can be removed. The PyTorch backend has it due to historical reasons, but for a new backend, it is not required.

Thanks for reply, this file seems not used in water.

**Comment by HydrogenSulfate** [comment]
> > Attention: Patch coverage is `55.25965%` with `2016 lines` in your changes missing coverage. Please review.
> 
> About 2000+ lines have yet to be tested in the CI. Could you take a look at the coverage report?



> > Attention: Patch coverage is `55.25965%` with `2016 lines` in your changes missing coverage. Please review.
> 
> About 2000+ lines have yet to be tested in the CI. Could you take a look at the coverage report?

Thanks, I will check it out and remove code not used in water.

**Comment by njzjz** [comment]
It seems to me that Paddle has a conflict cuDNN with JAX. It's hard to maintain the CUDA things for multiple software...

**Comment by HydrogenSulfate** [comment]
@njzjz hello, I found the issue lies with the nvidia-cuda-xxx series dependencies when the Paddle framework is installed, which have a unilateral impact on JAX. However, in the environment that JAX depends on, Paddle can works well. Therefore, my solution is to first install PaddlePaddle in the `test_cuda.yml` file, and then install JAX, Torch, and other packages. This way, the Paddle dependencies won't affect the other dependencies.

File: deepmd/pd/cxx_op.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    ----------
+    module_name : str
+        Name of the module
+
+    Returns
+    -------
+    bool
+        Whether the library is loaded successfully
+    """
+    # NOTE: Paddle do not support loading library from .so file yet.
```
Comment (njzjz) [inline deepmd/pd/cxx_op.py:17]: I am wondering whether this method will be used in the future? In fact, if no C++ codes are built, there is no need to change `backend` directory.

File: deepmd/pd/utils/cache.py
Comment (njzjz) [inline deepmd/pd/utils/cache.py:1]: This file can be remove if it is not used.

File: deepmd/pd/utils/learning_rate.py
Comment (njzjz) [inline deepmd/pd/utils/learning_rate.py:1]: This file can be removed and the redirected method can be used directly.

File: deepmd/pd/utils/plugin.py
Comment (njzjz) [inline deepmd/pd/utils/plugin.py:1]: This file can be removed and the redirected method can be used directly.

File: source/tests/pd/requirements.txt
Comment (njzjz) [inline source/tests/pd/requirements.txt:1]: This file can be removed.

**Review by njzjz — CHANGES_REQUESTED** [review]
I have some additional comments. We can merge this PR after v3.0.0 is released and the r3.0 branch is checked out.

**Comment by HydrogenSulfate** [comment]
> I have some additional comments. We can merge this PR after v3.0.0 is released and the r3.0 branch is checked out.

Thanks for the comments, redundant files&codes were removed.

**Review by njzjz — APPROVED** [review]

**Review by iProzd — APPROVED** [review]

**Comment by njzjz** [comment]
@coderabbitai resolve

**Comment by njzjz** [comment]
@coderabbitai resolve

File: deepmd/pd/infer/inference.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        """Construct a DeePMD tester.
+
+        Args:
+        - config: The Dict-like configuration with training options.
+        """
+        # Model
+        state_dict = paddle.load(model_ckpt)
+        if "model" in state_dict:
+            state_dict = state_dict["model"]
+        model_params = state_dict["_extra_state"]["model_params"]
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/pd/infer/inference.py:38]: @coderabbitai resolve

**Comment by njzjz** [comment]
> @coderabbitai resolve

... It seems to be not working. The comments may be required to be manually resolved

**Comment by HydrogenSulfate** [comment]
> @HydrogenSulfate A segfault appears in the GPU test. https://github.com/deepmodeling/deepmd-kit/actions/runs/12005580706/job/33462440790

I have make a new environment with cuda11.8 and installed the nightly build paddle(`3.0.0.dev20241126`), but all paddle tests running successfully as below.

``` shell
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Wed_Sep_21_10:33:58_PDT_2022
Cuda compilation tools, release 11.8, V11.8.89
Build cuda_11.8.r11.8/compiler.31833905_0
```

image

It seems a weird issue...

**Comment by njzjz** [comment]
> It seems a weird issue and is there any way to get the detailed error log from Test CUDA?

It's run in a docker container, so I think it may be reproducible.

Btw, the image is nvidia/cuda:12.6.2-cudnn-devel-ubuntu22.04

**Comment by njzjz** [comment]
A bit strange, but the segfault disappeared.

**Comment by HydrogenSulfate** [comment]
> > It seems a weird issue and is there any way to get the detailed error log from Test CUDA?
> 
> It's run in a docker container, so I think it may be reproducible.
> 
> Btw, the image is nvidia/cuda:12.6.2-cudnn-devel-ubuntu22.04

If we can add -v or -vv to pytest's argument so the test log can be a bit more detailed(but might not clear enough...)
